backend/rates/db.go

```diff
@@ -0,0 +1,64 @@
+package rates
+
+import (
+	"encoding/binary"
+	"math"
+	"path/filepath"
+	"time"
+
+	bbolt "github.com/coreos/bbolt"
+)
+
+func openRatesDB(dir string) (*bbolt.DB, error) {
+	opt := &bbolt.Options{Timeout: 5 * time.Second} // network disks may take long
+	return bbolt.Open(filepath.Join(dir, "rates.db"), 0600, opt)
+}
+
+// loadHistoryBucket loads data from an updater.historyDB bucket identified by the key.
+// The returned value is sorted by timestamp in ascending order.
+func (updater *RateUpdater) loadHistoryBucket(key string) ([]exchangeRate, error) {
+	var rates []exchangeRate
+	err := updater.historyDB.View(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket([]byte(key))
+		if bucket == nil {
+			return nil // no history exists for this key
+		}
+		return bucket.ForEach(func(k, v []byte) error {
+			timestamp := binary.LittleEndian.Uint64(k)
+			value := math.Float64frombits(binary.LittleEndian.Uint64(v))
+			rates = append(rates, exchangeRate{
+				value:     value,
+				timestamp: time.Unix(int64(timestamp), 0),
+			})
+			return nil
+		})
+	})
+	// Unfortunately, bbolt doesn't specify what kind of sorting order is used, if at all,
+	// in bucket.ForEach. Empirical testing confirms unidentified or no order,
+	// so we sort the slice in memory just to be sure.
+	// See TestReconfigureHistoryLoadsFromDB for details.
+	sortRatesByTimestamp(rates)
+	return rates, err
+}
+
+// dumpHistoryBucket stores rates in a DB bucket identified by the key.
+// It assumes rates are already sorted by timestamp in ascending order.
+func (updater *RateUpdater) dumpHistoryBucket(key string, rates []exchangeRate) error {
+	return updater.historyDB.Update(func(tx *bbolt.Tx) error {
+		tx.DeleteBucket([]byte(key)) //nolint:errcheck // don't care: next line will fail anyway
+		bucket, err := tx.CreateBucket([]byte(key))
+		if err != nil {
+			return err
+		}
+		for _, rate := range rates {
```

Comment:
what's the reason for storing each entry as an individual key:value row, instead of dumping all of it in serialized as one value? If you are loading and dumping all of them every time, there would be less overhead in storage and possibly performance?

Reply:
Reason: so that we have alternatives this dump/load version in the future. Maybe we'll find out in-memory map or its lock or something else is going to be a bottleneck and will want to switch to using bbold for the actual look ups in PriceAt.

Performance: it's an interesting assumption. To verify that, I wrote a simple benchmark where it dumps and loads data in 3 different ways: individual rows, JSON and Gob. Here are the results from my laptop with i7-8565U CPU at 1.8GHz:

```
$ cd backend/rates/
$ go test -bench=. -test.run=Benchmark
goos: linux
goarch: amd64
pkg: github.com/digitalbitbox/bitbox-wallet-app/backend/rates
BenchmarkDumpHistory/individual-8                    402           2912290 ns/op
BenchmarkDumpHistory/json-8                          282           4073714 ns/op
BenchmarkDumpHistory/gob-8                           938           1258093 ns/op
BenchmarkLoadHistory/individual-8                   4164            293393 ns/op
BenchmarkLoadHistory/json-8                          216           5476824 ns/op
BenchmarkLoadHistory/gob-8                          1716            695656 ns/op
PASS
ok      github.com/digitalbitbox/bitbox-wallet-app/backend/rates        8.675s
```

So, a single key with a JSON or Gob -encoded value is faster on writes (DumpHistory/xxx) but slower on reads (LoadHistory/xxx). This makes sense because:

> Bolt is good for read intensive workloads. Sequential write performance is also fast but random writes can be slow.

(https://github.com/etcd-io/bbolt#caveats--limitations)

In our case, we load the database once at startup (read) and dump new data every now and then from goroutines (write). We also load data in ReconfigureHistory in a goroutine from the backend handlers.

If we used one key with the whole bucket serialized as a value, if anything, it would most likely translate in slower startup and no gain during runtime.

Overhead in storage: I didn't include it in the benchmark but from my adhoc testing, 5000 of invidivual rows take about 1Mb of disk space, JSON is about 0.5Mb and Gob is 0.26Mb. Didn't try with protobuf but three versions I used already give good metrics to reason about imho. This storage overhead seems like a drop in the ocean compared to 100s of Mb of blockchain headers.

Reply:
wow, this answer goes above and beyond. nice and ACK!